lib/MonadicRewrite.thy

```diff
@@ -255,29 +326,117 @@ lemma monadic_rewrite_symb_exec_l'_preserve_names:
      \<not> F \<longrightarrow> no_fail P' m;
      \<And>rv. monadic_rewrite F E (Q rv) (x rv) y;
      \<lbrace>P\<rbrace> m \<lbrace>Q\<rbrace> \<rbrakk>
-      \<Longrightarrow> monadic_rewrite F E (P and P') (m >>= (\<lambda>rv. x rv)) y"
+   \<Longrightarrow> monadic_rewrite F E (P and P') (m >>= (\<lambda>rv. x rv)) y"
   by (rule monadic_rewrite_symb_exec_l')
 
 (* FIXME merge into below upon change-over desribed above *)
 lemmas monadic_rewrite_symb_exec_l'_TT
-    = monadic_rewrite_symb_exec_l'_preserve_names[where P'="\<top>" and F=True, simplified]
+  = monadic_rewrite_symb_exec_l'_preserve_names[where P'="\<top>" and F=True, simplified]
 
 lemmas monadic_rewrite_symb_exec_l
-    = monadic_rewrite_symb_exec_l''[where F=True and P'=\<top>, simplified]
-      monadic_rewrite_symb_exec_l''[where F=False, simplified simp_thms]
+  = monadic_rewrite_symb_exec_l''[where F=True and P'=\<top>, simplified]
+    monadic_rewrite_symb_exec_l''[where F=False, simplified simp_thms]
+
+lemma monadic_rewrite_symb_exec_l_known:
+  "\<lbrakk> \<And>s. \<lbrace>(=) s\<rbrace> m \<lbrace>\<lambda>r. (=) s\<rbrace>; empty_fail m;
+     monadic_rewrite True False Q (x rv) y;
+     \<lbrace>P\<rbrace> m \<lbrace>\<lambda>rv' s. rv' = rv \<and> Q s\<rbrace> \<rbrakk>
+   \<Longrightarrow> monadic_rewrite True False P (m >>= x) y"
+  apply (erule(1) monadic_rewrite_symb_exec_l)
+   apply (rule_tac P="rva = rv" in monadic_rewrite_gen_asm)
+   apply simp
+  apply wpsimp
+  done
+
+text \<open>if\<close>
+
+lemma monadic_rewrite_if:
+  "\<lbrakk> P \<Longrightarrow> monadic_rewrite F E Q a c; \<not> P \<Longrightarrow> monadic_rewrite F E R b d \<rbrakk>
+   \<Longrightarrow> monadic_rewrite F E (\<lambda>s. (P \<longrightarrow> Q s) \<and> (\<not> P \<longrightarrow> R s))
+                          (If P a b) (If P c d)"
+  by (cases P, simp_all)
+
+lemma monadic_rewrite_if_rhs:
+  "\<lbrakk> P \<Longrightarrow> monadic_rewrite F E Q a b; \<not> P \<Longrightarrow> monadic_rewrite F E R a c \<rbrakk>
+   \<Longrightarrow> monadic_rewrite F E (\<lambda>s. (P \<longrightarrow> Q s) \<and> (\<not> P \<longrightarrow> R s)) a (If P b c)"
+  by (cases P, simp_all)
+
+lemmas monadic_rewrite_named_if
+  = monadic_rewrite_if[where Q="(=) s" and R="(=) s", simplified] for s
+
+lemma monadic_rewrite_if_lhs:
+  "\<lbrakk> P \<Longrightarrow> monadic_rewrite F E Q b a; \<not> P \<Longrightarrow> monadic_rewrite F E R c a \<rbrakk>
+   \<Longrightarrow> monadic_rewrite F E (\<lambda>s. (P \<longrightarrow> Q s) \<and> (\<not> P \<longrightarrow> R s)) (If P b c) a"
+  by (cases P, simp_all)
+
+lemma monadic_rewrite_if_known:
+  "monadic_rewrite F E ((\<lambda>s. C = X) and \<top>) (if C then f else g) (if X then f else g)"
```

Comment:
Is the `and \<top>` doing anything here or can it be removed?

Reply:
I honestly have no idea. It can definitely be removed from the lemma, but as to how it is used in practice during monadic rewrite, I have not a clue.